pkgs/applications/networking/cluster/k3s/default.nix

```diff
@@ -260,27 +220,79 @@ stdenv.mkDerivation rec {
     conntrack-tools
   ];
 
-  buildInputs = [
-    k3sBin
-  ] ++ k3sRuntimeDeps;
+  buildInputs = k3sRuntimeDeps;
 
-  nativeBuildInputs = [ makeWrapper ];
+  nativeBuildInputs = [
+    makeWrapper
+    rsync
+    yq-go
+    zstd
+  ];
+
+  # embedded in the final k3s cli
+  propagatedBuildInputs = [
+    k3sCNIPlugins
+    k3sContainerd
+    k3sServer
+    runc
+  ];
 
-  unpackPhase = "true";
+  # Override the module build phase specifically to skip 'runHook preBuild'
+  # Why? Because in the 'preBuild' phase for the package, we run 'go generate'.
+  # The 'go generate' code runs 'go run ./pkg/...' stuff.
+  # That go generate code thus depends on having a module directory to go run
+  # stuff... so we can't run the go generate before the module vendoring step.
+  # But the go-module part of buildGoModule does just that, it runs our
```

Comment:
This might be something to start a separate discussion about.
It feels like there should be a way to, perhaps, set `modPreBuildPhase` to override `modBuildPhase` using `preBuildPhase`. Something like that.

Reply:
If you can remove `proxyVendor` running `go generate` during `preBuild` should work and won't need any workarounds.

Reply:
If I remove `proxyVendor`, I get the following:

```
+ go generate
go: inconsistent vendoring in /build/tmp.dri1VHvdKf:
... (omitted lines)
        mvdan.cc/unparam: is replaced in go.mod, but not marked as replaced in vendor/modules.txt

        To ignore the vendor directory, use -mod=readonly or -mod=mod.
        To sync the vendor directory, run:
                go mod vendor
```

The `go generate` here is weird because it does [this](https://github.com/k3s-io/k3s/blob/4f36c82ff771543af6d6010ce41287cdd2f659e7/main.go#L3)

```go
//go:generate go run pkg/codegen/main.go
```

The `go run` there needs access to the `go.mod`'s dependencies in order to build that package to run... and I can't figure out a way to make that work without proxyVendor being set.

It also doesn't seem like `go generate` can be correct during the `preBuild` of the `modBuildPhase`, because necessarily that preBuild is _before_ the module is available.

Do you have any idea for things I could try, or other ways to make this work?

Reply:
Could all of it be done in an overridden `buildPhase`?

Would mean needing to run `go build` and set `ldflags` and so forth but that way be simpler than trying to work around the vendor problem?

Reply:
I moved it all into a build phase, and I think it's a little better that way!

I'm still stuck with `proxyVendor`, or else the `go generate` step is unable to correctly find build dependencies for its internal `go run`... but overall, that doesn't seem like a big deal.